### PR TITLE
Add tests.

### PR DESCRIPTION
**Pull request by floitsch**

### CODE REVIEW

File: tests/broker_internal.toit [hunk 1]
```diff
@@ -0,0 +1,19 @@
+// Copyright (C) 2022 Toitware ApS.
+// Use of this source code is governed by a Zero-Clause BSD license that can
+// be found in the tests/LICENSE file.
+
+import log
+import mqtt.broker
+import .transport
+
+with_internal_broker --logger/log.Logger [block]:
+
```
Comment (kasperl) [inline tests/broker_internal.toit:10]: Drop newline.

File: tests/broker_mosquitto.toit [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+// Use of this source code is governed by a Zero-Clause BSD license that can
+// be found in the tests/LICENSE file.
+
+import host.pipe
+import log
+import monitor
+import mqtt
+import mqtt.transport as mqtt
+import net
+
```
Comment (kasperl) [inline tests/broker_mosquitto.toit:11]: Drop one newline.

File: tests/broker_mosquitto.toit [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+      ["mosquitto", "-v", "-p", port]  // Args.
+  return [
+    int.parse port,
+    fork_data
+  ]
+
+with_mosquitto --logger/log.Logger [block]:
+  mosquitto_data := start_mosquitto
+  port := mosquitto_data[0]
+  logger.info "Started Mosquitto on port $port"
```
Comment (kasperl) [inline tests/broker_mosquitto.toit:31]:
```suggestion
  logger.info "started mosquitto on port $port"
```

File: tests/close_test.toit [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+    // Since we connected, we also sent a disconnect packet.
+    expect_equals 1 (activity.filter: it[0] == "write" and it[1] is mqtt.DisconnectPacket).size
+
+close_in_handle create_transport/Lambda --logger/log.Logger --force/bool:
+  client := mqtt.FullClient --transport=create_transport.call --logger=logger
+
+  options := mqtt.SessionOptions --client_id="close_in_handle"
+  client.connect --options=options
+
+  handle_done := monitor.Latch
```
Comment (kasperl) [inline tests/close_test.toit:179]: It feels like being able to join a task would be useful.

File: tests/invalid_topic_test.toit [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+  callbacks := {:}
+  task::
+    exception := catch:
+      client.handle: | packet/mqtt.Packet |
+        if packet is mqtt.PublishPacket:
+          client.ack packet
+          publish := packet as mqtt.PublishPacket
+          callbacks[publish.topic].call publish
+        else:
+          logger.info "Ignored $(mqtt.Packet.debug_string_ packet)"
```
Comment (kasperl) [inline tests/invalid_topic_test.toit:48]:
```suggestion
          logger.info "ignored $(mqtt.Packet.debug_string_ packet)"
```

File: tests/full_client_test.toit [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+
+  callbacks := {:}
+  task::
+    client.handle: | packet/mqtt.Packet |
+      if packet is mqtt.PublishPacket:
+        client.ack packet
+        publish := packet as mqtt.PublishPacket
+        callbacks[publish.topic].call publish
+      else:
+        logger.info "Ignored $(mqtt.Packet.debug_string_ packet)"
```
Comment (kasperl) [inline tests/full_client_test.toit:128]:
```suggestion
        logger.info "ignored $(mqtt.Packet.debug_string_ packet)"
```

File: tests/full_client_test.toit [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+    client.unsubscribe response_topic
+    got_response.set true
+
+  client.publish topic "message".to_byte_array
+  got_response.get
+
+test_multisub client/mqtt.FullClient callbacks/Map --logger/log.Logger:
+  client.subscribe "idle"
+  2.repeat: | max_qos |
+    logger.info "Testing multi-subscription with max-qos=$max_qos"
```
Comment (kasperl) [inline tests/full_client_test.toit:82]:
```suggestion
    logger.info "testing multi-subscription with max-qos=$max_qos"
```

File: tests/full_client_test.toit [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+      subscription := ?
+      topic := ?
+      if sub_topic is string:
+        subscription = sub_topic
+        topic = sub_topic
+      else:
+        subscription = sub_topic[0]
+        topic = sub_topic[1]
+
+      logger.info "Testing topic: $topic - $subscription"
```
Comment (kasperl) [inline tests/full_client_test.toit:43]:
```suggestion
      logger.info "testing topic: $topic - $subscription"
```

File: tests/broker_mosquitto.toit [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+
+  mosquitto_is_running.get
+
+  network := net.open
+
+  try:
+    block.call:: mqtt.TcpTransport network --host="localhost" --port=port
+  finally: | is_exception _ |
+    pid := mosquitto_fork_data[3]
+    logger.info "Killing mosquitto server"
```
Comment (kasperl) [inline tests/broker_mosquitto.toit:60]:
```suggestion
    logger.info "killing mosquitto server"
```

File: tests/invalid_topic_test.toit [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+
+  exception := catch:
+    if mode == "subscribe":
+      client.subscribe topic
+    else if mode == "unsubscribe":
+      client.unsubscribe topic
+    else if mode == "publish":
+      client.publish topic #[]
+  if exception:
+    logger.info "Caught on client side"
```
Comment (kasperl) [inline tests/invalid_topic_test.toit:61]:
```suggestion
    logger.info "caught on client side"
```

File: tests/invalid_topic_test.toit [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+    client.close
+    return
+  else:
+    logger.info "Waiting for client to shut down"
```
Comment (kasperl) [inline tests/invalid_topic_test.toit:65]:
```suggestion
    logger.info "waiting for client to shut down"
```

File: tests/last_will_test.toit [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+  receiver_transport /mqtt.Transport := create_transport.call
+  receiver_client := mqtt.FullClient --transport=receiver_transport --logger=logger
+  receiver_options := mqtt.SessionOptions --client_id="receiver" --keep_alive=keep_alive --clean_session
+  receiver_client.connect --options=receiver_options
+
+  clients.do: | client |
+    task::
+      catch:
+        client.handle: | packet/mqtt.Packet |
+          logger.info "Received $(mqtt.Packet.debug_string_ packet)"
```
Comment (kasperl) [inline tests/last_will_test.toit:56]:
```suggestion
          logger.info "received $(mqtt.Packet.debug_string_ packet)"
```

File: tests/last_will_test.toit [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+
+        logger.info "client shut down"
+
+  subscription_ack_semaphore := monitor.Semaphore
+
+  received_packets := {:}
+  received_count := 0
+  task::
+    receiver_client.handle: | packet/mqtt.Packet |
+      logger.info "Received $(mqtt.Packet.debug_string_ packet)"
```
Comment (kasperl) [inline tests/last_will_test.toit:67]:
```suggestion
      logger.info "received $(mqtt.Packet.debug_string_ packet)"
```

File: tests/last_will_test.toit [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+
+  subscription_ack_semaphore.down
+  subscription_ack_semaphore.down
+  subscription_ack_semaphore.down
+  subscription_ack_semaphore.down
+
+  // Cut the connections of the clients, triggering a last will.
+  transports.do: it.close
+
+  while received_count != 4: sleep --ms=10
```
Comment (kasperl) [inline tests/last_will_test.toit:95]: Why not use a semaphore for this?

```
4.repeat: received.down
```

File: tests/last_will_test.toit [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+  received_count = 0
+  received_packets.clear
+
+  // Subscribe again to get the retained messages.
+  receiver_client.subscribe "test/last_will0"
+  receiver_client.subscribe "test/last_will1"
+  receiver_client.subscribe "test/last_will2"
+  receiver_client.subscribe "test/last_will3"
+
+  while received_count != 2: sleep --ms=10
```
Comment (kasperl) [inline tests/last_will_test.toit:117]: Ditto.

File: tests/multi_client_test.toit [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+  client2.connect --options=options2
+
+  client1_callback /Lambda := :: it // Ignore the packet.
+  client2_callback /Lambda := :: it // Ignore the packet.
+
+  2.repeat: | client_index |
+    client := client_index == 0 ? client1 : client2
+    task::
+      client.handle: | packet/mqtt.Packet |
+        logger.info "Received $(mqtt.Packet.debug_string_ packet)"
```
Comment (kasperl) [inline tests/multi_client_test.toit:44]:
```suggestion
        logger.info "received $(mqtt.Packet.debug_string_ packet)"
```

File: tests/packet_test_client.toit [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+
+  idle_topic := "idle-$client_id-$random"
+  wait_for_idle := ::
+    client.publish idle_topic #[] --qos=0
+    idle.down
+
+  task::
+    exception := catch --unwind=(on_handle_error == null):
+      client.handle: | packet/mqtt.Packet |
+        logger.info "Received $(mqtt.Packet.debug_string_ packet)"
```
Comment (kasperl) [inline tests/packet_test_client.toit:50]:
```suggestion
        logger.info "received $(mqtt.Packet.debug_string_ packet)"
```

File: tests/qos_test.toit [hunk 18]
```diff
@@ -0,0 +1,161 @@
...
+
+  wait_for_idle := :
+    client.publish "idle" #[] --qos=0
+    idle.down
+
+  auto_ack := true
+
+  task::
+    client.handle: | packet/mqtt.Packet |
+      logger.info "Received $(mqtt.Packet.debug_string_ packet)"
```
Comment (kasperl) [inline tests/qos_test.toit:130]:
```suggestion
      logger.info "received $(mqtt.Packet.debug_string_ packet)"
```

File: tests/third_party/ephemeral-port-reserve/README.toitware [hunk 19]
```diff
@@ -0,0 +1,12 @@
...
+URL: https://github.com/Yelp/ephemeral-port-reserve
+Version: 1.1.4
+Revision: 6cf5addc2f1d5c0a25be4ba4af62104da1d8fc51
+Date: 2022-05-30
+Author: Yelp
+License: MIT
+
+Description:
+Reserves a port by forcing the port into a TIME_WAIT state which
+  gives subprocesses 60 seconds time to use the port.
```
Comment (kasperl) [inline tests/third_party/ephemeral-port-reserve/README.toitware:12]:
```suggestion
  gives subprocesses 60 seconds to use the port.
```

File: tests/transport.toit [hunk 20]
```diff
@@ -0,0 +1,276 @@
...
+  write bytes/ByteArray -> int:
+    pipe_.broker_write bytes
+    return bytes.size
+
+  read -> ByteArray?:
+    return pipe_.broker_read
+
+  close -> none:
+    pipe_.broker_close
+
```
Comment (kasperl) [inline tests/transport.toit:56]: Drop one newline.

File: tests/transport.toit [hunk 21]
```diff
@@ -0,0 +1,276 @@
...
+      throw "Transport is closed"
+
+    pipe := TestTransportPipe
+    channel_.send pipe
+    return pipe
+
+  close -> none:
+    is_closed = true
+    channel_.send null
+
```
Comment (kasperl) [inline tests/transport.toit:78]: Drop one newline.

File: tests/transport.toit [hunk 22]
```diff
@@ -0,0 +1,276 @@
...
+    data_ = null
+    return result
+
+  close:
+    is_closed_ = true
+
+  write bytes/ByteArray -> none:
+    await: not data_
+    data_ = bytes
+
```
Comment (kasperl) [inline tests/transport.toit:141]: Drop one newline.

File: tests/transport.toit [hunk 23]
```diff
@@ -0,0 +1,276 @@
...
+  wrapped_ /reader.Reader
+  intercepted /Deque := Deque
+
+  constructor .wrapped_:
+
+  read -> ByteArray?:
+    bytes := wrapped_.read
+    intercepted.add bytes
+    return bytes
+
```
Comment (kasperl) [inline tests/transport.toit:153]: Drop one newline.

**Review by kasperl — APPROVED** [review]
Looks good to me.

File: tests/broker_mosquitto.toit (hunk 2 above)
Comment (floitsch) [reply tests/broker_mosquitto.toit:11]: done.

File: tests/last_will_test.toit (hunk 15 above)
Comment (floitsch) [reply tests/last_will_test.toit]: done.

File: tests/transport.toit (hunk 22 above)
Comment (floitsch) [reply tests/transport.toit:139]: done.

File: tests/transport.toit (hunk 23 above)
Comment (floitsch) [reply tests/transport.toit:150]: done.

**Pull request merged**